e2e: TestCrossLogicalClusterList fails when run more than once against persistent fixture

The [TestCrossLogicalClusterList e2e test](https://github.com/kcp-dev/kcp/blob/main/test/e2e/conformance/cross_logical_cluster_list_test.go) fails if run more than once against a persistent server. This is curious, since the testing appears to be isolated within test-local organizations and workspaces.

**To Reproduce**
1. Start a persistent test server in one terminal:
```
$ rm -rf .kcp && make build && ./bin/test-server 2>&1 | tee kcp.log
```
2. Run the test in another terminal (it should pass):
```
 go test -v  test/e2e/conformance/cross_logical_cluster_list_test.go -args --kubeconfig=$(pwd)/.kcp/admin.kubeconfig
```
3. Run the same test again and it should quickly fail.

Copying failure output from https://github.com/kcp-dev/kcp/issues/796#issuecomment-1083391697 so we have it here:

```
=== RUN   TestCrossLogicalClusterList
=== PAUSE TestCrossLogicalClusterList
=== RUN   TestCrossLogicalClusterListPartialObjectMetadata
=== PAUSE TestCrossLogicalClusterListPartialObjectMetadata
=== CONT  TestCrossLogicalClusterList
=== CONT  TestCrossLogicalClusterListPartialObjectMetadata
=== CONT  TestCrossLogicalClusterList
    fixture.go:76: shared kcp server will target configuration /home/maru/src/kcp/main/.kcp/admin.kubeconfig
=== CONT  TestCrossLogicalClusterListPartialObjectMetadata
    fixture.go:76: shared kcp server will target configuration /home/maru/src/kcp/main/.kcp/admin.kubeconfig
=== CONT  TestCrossLogicalClusterList
    cross_logical_cluster_list_test.go:73: Bootstrapping ClusterWorkspace CRDs in logical cluster root:e2e-org-wp5xc
I0330 09:58:32.170918  356010 bootstrap.go:126] Bootstrapped CRD root:e2e-org-wp5xc|clusterworkspaces.tenancy.kcp.dev after 5.408786ms
    cross_logical_cluster_list_test.go:87: Creating ClusterWorkspace CRs in logical cluster root:e2e-org-wp5xc
    util.go:117: Saving test artifacts for test "TestCrossLogicalClusterList" under "/tmp/TestCrossLogicalClusterList1202823195/001/TestCrossLogicalClusterList/40287696".
    cross_logical_cluster_list_test.go:73: Bootstrapping ClusterWorkspace CRDs in logical cluster root:e2e-org-zhpqc
I0330 09:58:32.294764  356010 bootstrap.go:126] Bootstrapped CRD root:e2e-org-zhpqc|clusterworkspaces.tenancy.kcp.dev after 16.058365ms
=== CONT  TestCrossLogicalClusterListPartialObjectMetadata
    cross_logical_cluster_list_test.go:143: Install a normal sheriffs CRD into workspace "root:e2e-org-6pwhp:e2e-workspace-phlxp"
I0330 09:58:32.381952  356010 bootstrap.go:126] Bootstrapped CRD root:e2e-org-6pwhp:e2e-workspace-phlxp|sheriffs.wildwest.dev after 3.130769ms
=== CONT  TestCrossLogicalClusterList
    cross_logical_cluster_list_test.go:87: Creating ClusterWorkspace CRs in logical cluster root:e2e-org-zhpqc
    cross_logical_cluster_list_test.go:104: Listing ClusterWorkspace CRs across logical clusters
    cross_logical_cluster_list_test.go:111: Expecting at least those ClusterWorkspaces we created above
    util.go:211: saving artifact to /tmp/TestCrossLogicalClusterList1202823195/001/TestCrossLogicalClusterList/40287696/artifacts/kcp/shared/root_e2e-org-zhpqc/clusterworkspaces.tenancy.kcp.dev_ws-1.yaml
    util.go:211: saving artifact to /tmp/TestCrossLogicalClusterList1202823195/001/TestCrossLogicalClusterList/40287696/artifacts/kcp/shared/root_e2e-org-wp5xc/clusterworkspaces.tenancy.kcp.dev_ws-0.yaml
--- PASS: TestCrossLogicalClusterList (0.50s)
=== CONT  TestCrossLogicalClusterListPartialObjectMetadata
    cross_logical_cluster_list_test.go:146: Install another normal sheriffs CRD into workspace "root:e2e-org-6pwhp:e2e-workspace-qnwmz"
I0330 09:58:32.488850  356010 bootstrap.go:126] Bootstrapped CRD root:e2e-org-6pwhp:e2e-workspace-qnwmz|sheriffs.wildwest.dev after 2.635875ms
    cross_logical_cluster_list_test.go:149: Trying to wildcard list
    cross_logical_cluster_list_test.go:151:
            Error Trace:    cross_logical_cluster_list_test.go:151
            Error:          Received unexpected error:
                            Internal error occurred: error resolving resource: Internal error occurred: error resolving resource: cannot watch across logical clusters for a resource type with several distinct schemas
            Test:           TestCrossLogicalClusterListPartialObjectMetadata
            Messages:       expected wildcard list to work because schemas are the same
--- FAIL: TestCrossLogicalClusterListPartialObjectMetadata (0.65s)
```

--- a/.github/workflows/ci.yaml
+++ b/.github/workflows/ci.yaml
@@ -143,7 +143,7 @@ jobs:
           ARTIFACT_DIR=/tmp/e2e \
           PATH="${PATH}:$(pwd)/bin/" \
           TEST_ARGS='-args --use-default-server' \
-          COUNT=2 \
+          COUNT=1 \
           E2E_PARALLELISM=2 \
           make test-e2e
       - uses: actions/upload-artifact@v2

--- a/test/e2e/conformance/cross_logical_cluster_list_test.go
+++ b/test/e2e/conformance/cross_logical_cluster_list_test.go
@@ -53,7 +53,8 @@ import (
 func TestCrossLogicalClusterList(t *testing.T) {
 	t.Parallel()
 
-	server := framework.SharedKcpServer(t)
+	// This test changes global state and is not compatible with shared fixture.
+	server := framework.PrivateKcpServer(t)
 
 	ctx, cancelFunc := context.WithCancel(context.Background())
 	t.Cleanup(cancelFunc)
@@ -120,7 +121,8 @@ func TestCrossLogicalClusterListPartialObjectMetadata(t *testing.T) {
 	// ensure PartialObjectMetadata wildcard list works even with different CRD schemas
 	t.Parallel()
 
-	server := framework.SharedKcpServer(t)
+	// This test changes global state and is not compatible with shared fixture.
+	server := framework.PrivateKcpServer(t)
 
 	ctx, cancelFunc := context.WithCancel(context.Background())
 	t.Cleanup(cancelFunc)

--- a/test/e2e/framework/fixture.go
+++ b/test/e2e/framework/fixture.go
@@ -58,8 +58,21 @@ func TestServerArgsWithTokenAuthFile(tokenAuthFile string) []string {
 }
 
 // KcpFixture manages the lifecycle of a set of kcp servers.
-type KcpFixture struct {
+//
+// Deprecated for use outside this package. Prefer PrivateKcpServer().
+type kcpFixture struct {
 	Servers map[string]RunningServer
+}
+
+// PrivateKcpServer returns a new kcp server fixture managing a new
+// server process that is not intended to be shared between tests.
+func PrivateKcpServer(t *testing.T, args ...string) RunningServer {
+	serverName := "main"
+	f := newKcpFixture(t, kcpConfig{
+		Name: serverName,
+		Args: args,
+	})
+	return f.Servers[serverName]
 }
 
 // SharedKcpServer returns a kcp server fixture intended to be shared
@@ -87,15 +100,16 @@ func SharedKcpServer(t *testing.T) RunningServer {
 	// fixture.
 
 	tokenAuthFile := WriteTokenAuthFile(t)
-	f := NewKcpFixture(t, KcpConfig{
+	f := newKcpFixture(t, kcpConfig{
 		Name: serverName,
 		Args: TestServerArgsWithTokenAuthFile(tokenAuthFile),
 	})
 	return f.Servers[serverName]
 }
 
-func NewKcpFixture(t *testing.T, cfgs ...KcpConfig) *KcpFixture {
-	f := &KcpFixture{}
+// Deprecated for use outside this package. Prefer PrivateKcpServer().
+func newKcpFixture(t *testing.T, cfgs ...kcpConfig) *kcpFixture {
+	f := &kcpFixture{}
 
 	artifactDir, dataDir, err := ScratchDirs(t)
 	require.NoError(t, err, "failed to create scratch dirs: %v", err)

--- a/test/e2e/framework/kcp.go
+++ b/test/e2e/framework/kcp.go
@@ -70,7 +70,7 @@ type kcpServer struct {
 	t *testing.T
 }
 
-func newKcpServer(t *testing.T, cfg KcpConfig, artifactDir, dataDir string) (*kcpServer, error) {
+func newKcpServer(t *testing.T, cfg kcpConfig, artifactDir, dataDir string) (*kcpServer, error) {
 	t.Helper()
 
 	kcpListenPort, err := GetFreePort(t)

--- a/test/e2e/framework/test.go
+++ b/test/e2e/framework/test.go
@@ -32,8 +32,10 @@ type RunningServer interface {
 	Artifact(t *testing.T, producer func() (runtime.Object, error))
 }
 
-// KcpConfig qualify a kcp server to start
-type KcpConfig struct {
+// kcpConfig qualify a kcp server to start
+//
+// Deprecated for use outside this package. Prefer PrivateKcpServer().
+type kcpConfig struct {
 	Name string
 	Args []string
 

--- a/test/e2e/reconciler/workspace/controller_test.go
+++ b/test/e2e/reconciler/workspace/controller_test.go
@@ -191,9 +191,7 @@ func TestWorkspaceController(t *testing.T) {
 				// Destructive tests require their own server
 				//
 				// TODO(marun) Could the testing currently requiring destructive e2e be performed with less cost?
-				const serverName = "main"
-				f := framework.NewKcpFixture(t, framework.KcpConfig{Name: serverName})
-				server = f.Servers[serverName]
+				server = framework.PrivateKcpServer(t)
 			}
 
 			cfg := server.DefaultConfig(t)

--- a/test/e2e/reconciler/workspaceshard/controller_test.go
+++ b/test/e2e/reconciler/workspaceshard/controller_test.go
@@ -323,9 +323,7 @@ func TestWorkspaceShardController(t *testing.T) {
 				// Destructive tests require their own server
 				//
 				// TODO(marun) Could the testing currently requiring destructive e2e be performed with less cost?
-				const serverName = "main"
-				f := framework.NewKcpFixture(t, framework.KcpConfig{Name: serverName})
-				server = f.Servers[serverName]
+				server = framework.PrivateKcpServer(t)
 			}
 
 			cfg := server.DefaultConfig(t)

--- a/test/e2e/virtual/workspaces/virtual_workspace_test.go
+++ b/test/e2e/virtual/workspaces/virtual_workspace_test.go
@@ -519,21 +519,14 @@ func testWorkspacesVirtualWorkspaces(t *testing.T, standalone bool) {
 		portStr, err := framework.GetFreePort(t)
 		require.NoError(t, err)
 
-		serverName := "main"
 		tokenAuthFile := framework.WriteTokenAuthFile(t)
-		f := framework.NewKcpFixture(t,
-			framework.KcpConfig{
-				Name: serverName,
-				Args: []string{
-					"--run-controllers=false",
-					"--unsupported-run-individual-controllers=workspace-scheduler",
-					"--run-virtual-workspaces=false",
-					fmt.Sprintf("--virtual-workspace-address=https://localhost:%s", portStr),
-					"--token-auth-file", tokenAuthFile,
-				},
-			},
+		server = framework.PrivateKcpServer(t,
+			"--run-controllers=false",
+			"--unsupported-run-individual-controllers=workspace-scheduler",
+			"--run-virtual-workspaces=false",
+			fmt.Sprintf("--virtual-workspace-address=https://localhost:%s", portStr),
+			"--token-auth-file", tokenAuthFile,
 		)
-		server = f.Servers[serverName]
 
 		// write kubeconfig to disk, next to kcp kubeconfig
 		kcpAdminConfig, _ := server.RawConfig()

